### ISSUE TITLE
race condition

### ISSUE THREAD

**Issue by aveeday** [dev-mastery/comments-api#37]
Next code from use case add-comment has race condition error in design.

If two the same requests will be received at the same time, 
then `if (exists)` will be false for each of them
and two comments with same hash will be inserted

```js
...
    const exists = await commentsDb.findByHash({ hash: comment.getHash() })
    if (exists) {
      return exists
    }

    const moderated = await handleModeration({ comment })
    const commentSource = moderated.getSource()
    return commentsDb.insert({
...
```

This example app has no business rules.
Crud methods can be mapped to controller without need of use cases.
If I need to add real business logic this error will be critical e.g:
- user should not be able to post more than 1 comment per hour

How can I avoid this error in clean architecture?

**Comment by arcdev1** [comment]
Good catch. There is a tiny window where simultaneous requests would result in a duplicate record. The most straightforward solution is to add a `unique` constraint to the `hash` field in the database and handle the resulting exception appropriately.
